pkg/sqlutil/pg/pg.go

```diff
@@ -1,35 +1,132 @@
+	if driverName == DriverTxWrappedPostgres {
+		// txdb uses the uri as a unique identifier for each transaction. Each ORM should
+		// be encapsulated in its own transaction, and thus needs its own unique id.
+		//
+		// We can happily throw away the original uri here because if we are using txdb
+		// it should have already been set at the point where we called [txdb.Register].
+		return config.open(ctx, uuid.NewString(), driverName)
+	}
+	return config.openDB(ctx, uri, driverName)
+}
+
+func (config DBConfig) open(ctx context.Context, uri string, driverName string) (*sqlx.DB, error) {
+	sqlDB, err := otelsql.Open(driverName, uri, otelOpts...)
 	if err != nil {
-		t.Fatalf("failed to register txdb dialect: %s", err.Error())
-		return nil
+		return nil, err
+	}
+	if _, err := sqlDB.ExecContext(ctx, config.initSQL()); err != nil {
+		return nil, err
 	}
-	db, err := sqlx.Open(string(TransactionWrappedPostgres), uuid.New().String())
-	require.NoError(t, err)
-	t.Cleanup(func() { assert.NoError(t, db.Close()) })
-	db.MapperFunc(reflectx.CamelToSnakeASCII)
 
-	return db
+	return config.newDB(ctx, sqlDB, driverName)
 }
 
-func TestURL(t testing.TB) string {
-	dbURL, ok := os.LookupEnv("CL_DATABASE_URL")
-	if ok {
-		return dbURL
+func (config DBConfig) openDB(ctx context.Context, uri string, driverName string) (*sqlx.DB, error) {
+	connConfig, err := pgx.ParseConfig(uri)
+	if err != nil {
+		return nil, fmt.Errorf("database: failed to parse config: %w", err)
 	}
-	t.Log("CL_DATABASE_URL not set--falling back to testing txdb backed by an in-memory db")
-	return string(InMemoryPostgres)
+
+	initSQL := config.initSQL()
+	connector := stdlib.GetConnector(*connConfig, stdlib.OptionAfterConnect(func(ctx context.Context, c *pgx.Conn) (err error) {
+		_, err = c.Exec(ctx, initSQL)
+		return
+	}))
+
+	sqlDB := otelsql.OpenDB(connector, otelOpts...)
+	return config.newDB(ctx, sqlDB, driverName)
 }
+
+func (config DBConfig) initSQL() string {
+	return fmt.Sprintf(`SET TIME ZONE 'UTC'; SET lock_timeout = %d; SET idle_in_transaction_session_timeout = %d; SET default_transaction_isolation = %q`,
+		config.LockTimeout.Milliseconds(), config.IdleInTxSessionTimeout.Milliseconds(), defaultIsolation)
+}
+
+func (config DBConfig) newDB(ctx context.Context, sqldb *sql.DB, driverName string) (*sqlx.DB, error) {
+	db := sqlx.NewDb(sqldb, driverName)
+	db.MapperFunc(reflectx.CamelToSnakeASCII)
+	db.SetMaxOpenConns(config.MaxOpenConns)
+	db.SetMaxIdleConns(config.MaxIdleConns)
+	return db, disallowReplica(ctx, db)
+}
+
+func disallowReplica(ctx context.Context, db *sqlx.DB) error {
+	var val string
+	err := db.GetContext(ctx, &val, "SHOW session_replication_role")
+	if err != nil {
+		return err
+	}
+
+	if val == "replica" {
+		return fmt.Errorf("invalid `session_replication_role`: %s. Refusing to connect to replica database. Writing to a replica will corrupt the database", val)
+	}
+
+	return nil
+}
+
+// Deprecated: Use Driver* constants
+type DialectName string
```

Comment:
This type was forcing type conversions and wasn't restricting what could be used, so it is cleaner to use plain string instead. Leaving it here temporarily for compatibility though.

Reply:
(also it seems to be called 'driver name' everywhere? not sure where the term 'dialect' orginated from)